containers-tests/tests/intmap-strictness.hs

```diff
@@ -76,6 +82,70 @@ pInsertLookupWithKeyValueStrict f k v m
                      not (isBottom $ M.insertLookupWithKey (const3 1) k bottom m)
     | otherwise    = isBottom $ M.insertLookupWithKey (apply3 f) k bottom m
 
+pFromAscListLazy :: [Int] -> Bool
+pFromAscListLazy ks = not . isBottom $ M.fromAscList elems
+  where
+    ks' = map head . group . sort $ ks
+    elems = [(k, v) | k <- ks', v <- [undefined, ()]]
+
+pFromAscListStrict :: [Int] -> Bool
+pFromAscListStrict ks
+    | null ks   = not . isBottom $ M.fromAscList elems
+    | otherwise = isBottom $ M.fromList elems
+  where
+    ks' = map head . group . sort $ ks
+    elems = [(k, v) | k <- ks', v <- [undefined, undefined, ()]]
+
+------------------------------------------------------------------------
+-- check for extra thunks
+
+tExtraThunksM :: Test
+tExtraThunksM = testGroup "IntMap.Strict - extra thunks" $
+    if not isUnitSupported then [] else
+    -- for strict maps, all the values should be evaluated to ()
+    [ check "singleton"       True  $ m0
+    , check "insert"          True  $ M.insert 42 () m0
+    , check "insertWith"      True  $ M.insertWith const 42 () m0
+    , check "fromList"        True  $ M.fromList [(42,()),(42,())]
+    , check "fromListWith"    True  $ M.fromListWith const [(42,()),(42,())]
+    , check "fromAscList"     True  $ M.fromAscList [(42,()),(42,())]
+    , check "fromAscListWith" True  $ M.fromAscListWith const [(42,()),(42,())]
+    , check "fromDistinctAscList" True $ M.fromAscList [(42,())]
+    ]
+  where
+    m0 = M.singleton 42 ()
+    check :: TestName -> Bool -> IntMap () -> Test
+    check n e m = testCase n $ case M.lookup 42 m of
+        Just v -> assertBool msg (e == isUnit v)
+        _      -> assertString "key not found"
+      where
+        msg | e         = "too lazy -- expected fully evaluated ()"
+            | otherwise = "too strict -- expected a thunk"
```

Comment:
Looks like we never use `check ... False` here, so this could possibly be simplified?!